File: problems/0200-0300/0210_course-schedule-ii/solution_kahn.cpp
class Solution {
  public:
    vector<int> findOrder(int numCourses, vector<vector<int>> &prerequisites) {
        // A valid order is exactly a topological ordering of the graph where
        // each pair [course, prereq] is the edge prereq -> course.
        vector<vector<int>> adjacency(numCourses);
        vector<int> indegree(numCourses, 0);
        for (const auto &pair : prerequisites) {
            int course = pair[0];
            int prereq = pair[1];
            adjacency[prereq].push_back(course);
            indegree[course] += 1;
        }
        // Kahn's algorithm: start from every course with no prerequisites.
        queue<int> q;
        for (int i = 0; i < numCourses; i++) {
            if (indegree[i] == 0) {
                q.push(i);
            }
        }
        vector<int> order;
        order.reserve(numCourses);
        while (!q.empty()) {
            int node = q.front();
            q.pop();
            order.push_back(node);
            // Emitting a course consumes its edges: dependents lose one
            // prerequisite, and any that reaches zero becomes available.
            for (int nxt : adjacency[node]) {
                if (--indegree[nxt] == 0) {
                    q.push(nxt);
                }
            }
        }
        // A shortfall means a cycle kept positive indegrees forever; the
        // problem requires an empty list rather than a partial order.
        if ((int)order.size() == numCourses) {
            return order;
        }
        return {};
    }
};

File: problems-bettercode/0201-0300/0210_course-schedule-ii/solution_kahn.go
func findOrder(numCourses int, prerequisites [][]int) []int {
	// A valid order is exactly a topological ordering of the graph where
	// each pair [course, prereq] is the edge prereq -> course.
	adjacency := make([][]int, numCourses)
	indegree := make([]int, numCourses)
	for _, pair := range prerequisites {
		course, prereq := pair[0], pair[1]
		adjacency[prereq] = append(adjacency[prereq], course)
		indegree[course]++
	}
	// Kahn's algorithm: start from every course with no prerequisites.
	queue := make([]int, 0, numCourses)
	for i := 0; i < numCourses; i++ {
		if indegree[i] == 0 {
			queue = append(queue, i)
		}
	}
	order := make([]int, 0, numCourses)
	for head := 0; head < len(queue); head++ {
		node := queue[head]
		order = append(order, node)
		// Emitting a course consumes its edges: dependents lose one
		// prerequisite, and any that reaches zero becomes available.
		for _, nxt := range adjacency[node] {
			indegree[nxt]--
			if indegree[nxt] == 0 {
				queue = append(queue, nxt)
			}
		}
	}
	// A shortfall means a cycle kept positive indegrees forever; the problem
	// requires an empty list rather than a partial order.
	if len(order) == numCourses {
		return order
	}
	return []int{}
}
